### ISSUE TITLE
Replace the Azure's last_dates.json file with a database file and an ORM

### ISSUE THREAD

**Issue by CarlosRS9** [wazuh/wazuh#11940]
| Affected integration |
|---|
| Azure |

## Description

The current implementation of the Azure integration uses a JSON file to store relevant information about the previous executions, used to avoid getting duplicated logs when working with different `offset` values. This is nos consistent with the AWS or GCP integrations which uses a database instead.

Using a plain text file, even if it is in JSON format, is counterproductive because the larger the file the more inefficient the module will be, since the module needs to load it in memory, operate with it and overwrite it on different occasions. This is also a problem when performing operations that are trivial when dealing with a real database. For these reasons, we must replace this file with a real database.

To tackle this task it would be advisable to implement a simple ORM, similar to the one that exists in the framework, which allows us on the one hand to correctly manage the migration of the JSON file to a real database and on the other hand to facilitate the access to the information stored during the execution of the module.

## Tasks
- [x] Apply the changes.
- [x] Test in a manager.
- [x] Test in an agent.
- [x] Unit tests without failures. Updated if there are any relevant changes.
- [x] API integration tests without failures. Updated if there are any relevant changes.
- [x] Update the documentation if necessary.
- [x] Add entry to the changelog if necessary.

**Comment by CarlosRS9** [comment]
# Status update

We are currently working on the ORM implementation. Here are some details about its design:

- It makes use of `SQLAlchemy`, similar to how the RBAC ORM works in the Framework.

- The database will consist of 3 tables, one for each service: `Graph`, `LogAnalytics` and `Storage`.

- The three tables will have the same structure:
	- `md5`: the md5 value of the query to process, used as a key. The md5 has been kept to avoid breaking compatibility with the old module when migrating the last_dates.json file.
	- query`: the text string of the query to execute. It can be used in future versions to replace the `md5` field if necessary. Until then, it will serve as additional debug info when debugging the module.
	- `min`: The smallest date processed. This already existed in `last_dates.json`.
	- `max`: The highest date processed. This already existed in `last_dates.json`.

- When the module runs, a check is made to determine if there is a `last_dates.json` file to migrate.

- Here is the functionality that the ORM will provide:
	- check if database exists
	- check if last_dates file exists
	- create database file
	- migrate data from last_dates
	- connect with the database
	- close connection with the database
	- Insert data into table
	- Read data from the different tables available

### Considerations

As mentioned, the use of `md5` as a key to identify a query in the database has been retained. This is to maintain compatibility or else duplicate logs could be processed during the first run after a migration.

### Pending

- Finish implementing changes
- Determine whether the tests will be implemented in this issue or in the UT epic.

**Comment by CarlosRS9** [comment]
# Status update

We are working in parallel on both the tests and the main logic of the ORM module, as well as other minor changes needed. Development is well underway. The ORM tests are almost ready.

Here are the new files that will be part of the Azure module structure:

- `azure/orm.py`: Contains all the functionality needed to manage the database, as well as to maintain its integrity when updating.
- `azure/tools.py`: Required tools common to `azure-logs.py` or `orm.py`. For example, functions such as loading the json file into memory and processing it.
- `azure/tests/test_orm.py`: Test suite for the ORM.
- `azure/tests/test/test_tools.py`: Suite of tests for the tools.
- `azure/tests/data/last_date_files/:` Files needed for testing the ORM module.

As the Azure module did not have any unit tests we are still making the necessary changes to add them. In this issue we will only add the tests related to the introduced changes, i.e. `test_orm` and `test_tools`.

In addition to this, the `azure-logs.py` file is being modified to adapt it to the new ORM usage.

**Comment by CarlosRS9** [comment]
# Status update

We have finished implementing all the unit tests for the newly added `tools.py` and `orm.py` modules. We have finished applying the necessary changes to the Azure module to make use of the new ORM functionality.

We have started manual testing of the module. During these tests we have noticed a duplication problem derived from Azure's use of datetime to indicate the activity date for the events. As we pointed out in the original Azure issue, in some cases the Azure API returns a value for the activity date whose value for microseconds has 7 decimal places. This is incompatible with the python datetime standard, which requires 6 digits. The database had been created to store Datetime objects, so it was missing this last digit, losing precision which in the end translates into the possibility of duplicate events. To solve this we are changing the database to store these dates in string format, in order to store them as they are returned by the Azure API.

**Comment by CarlosRS9** [comment]
# Regarding `tools.py` and the wodles custom python package

Both the gcloud module and now the Azure module had the same file called `tools.py`. This resulted in several problems related to the way in which the different imports are performed. The solution proposed in #12874 was to create a custom python package, similar to what we have for the framework or the api, that includes all the modules contained in the `wodles/` folder.

Due to the impact that creating a custom python package would have on the other components such as the integration tests and the other repositories, as well as the low priority of this change, it was decided to discard it for the moment. 

Finally, it was decided to remove the conflicting `tools.py` file from Azure as it only contained the `load_last_dates_file` function, which was only used in the `orm.py` module. We have opted to move this function to the `orm.py` and update all tests accordingly.

Therefore, we have removed:
- `tools.py`
- `test_tools.py`
